Bug 2094196: Added webhook support for PowerVS platform

Karthik-K-N opened the pull request:
This PR will add validation and defaulting webhook support for PowerVS Platform

Sample Machine Spec can be found [here](https://github.com/openshift/machine-api-provider-powervs/blob/81e0c5c8fc2dc5b2f5ff2e32a37ff673ac9ba407/examples/machine-with-user-data.yaml)

Karthik-K-N commented:
/cc @mkumatag

mkumatag commented on pkg/webhooks/machine_webhook.go:
```diff
@@ -110,6 +111,13 @@ const (
+	defaultPowerVSCredentialsSecret = "powervs-credentials"
+	defaultPowerVSSysType           = "s922"
+	defaultPowerVSProcType          = "shared"
+	defaultPowerVSProcessor         = "0.25"
+	defaultPowerVSMemory            = "8"
```
@clnperez need your help deciding what should be the right values here for any worker ^^

Karthik-K-N replied:
@clnperez shall I go ahead with this https://github.com/openshift-powervs/installer/blob/2e17e5afefda5976366d71caa7cdee03fa42ceec/pkg/asset/machines/worker.go#L165  as default values.

clnperez replied:
i would recommend the values the UPI folks have been using: https://github.com/ocp-power-automation/ocp4-upi-powervs/blob/7279803eed6566faf820bcac1334dc8e7f5ba761/compute-vars/medium.tfvars#L6

mkumatag commented on go.mod:
```diff
@@ -32,6 +32,8 @@ require (
 	sigs.k8s.io/yaml v1.3.0
 )
 
+require github.com/openshift/machine-api-provider-powervs v0.0.0-20220303154846-89bb5bec120e
```
why a separate entry here? can we include in the above block itself?

Karthik-K-N replied:
Updated it in the latest commit

mkumatag commented on pkg/webhooks/machine_webhook.go:
```diff
@@ -110,6 +111,13 @@ const (
+	defaultPowerVSCredentialsSecret = "powervs-credentials"
+	defaultPowerVSSysType           = "s922"
+	defaultPowerVSProcType          = "shared"
+	defaultPowerVSProcessor         = "0.25"
```
```suggestion
	defaultPowerVSProcessor         = "0.5"
```

Karthik-K-N commented:
/test unit

mkumatag commented on pkg/webhooks/machineset_webhook_test.go:
```diff
@@ -216,22 +225,107 @@ func TestMachineSetCreation(t *testing.T) {
+			clusterID:    "powervs-cluster",
 			providerSpecValue: &runtime.RawExtension{
-				Object: &machinev1.VSphereMachineProviderSpec{
-					Template: "template",
-					Workspace: &machinev1.Workspace{
-						Datacenter: "datacenter",
-						Server:     "server",
+				Object: &powervsmachine.PowerVSMachineProviderConfig{
+					ServiceInstanceID: "TestServiceInstanceID",
+					KeyPairName:       "TestKeyPair",
+					Image: powervsmachine.PowerVSResourceReference{
+						Name: pointer.StringPtr("TestNetworkName"),
 					},
-					Network: machinev1.NetworkSpec{
-						Devices: []machinev1.NetworkDeviceSpec{
-							{
-								NetworkName: "networkName",
-							},
-						},
```
why are we modifying the existing testcase?

Karthik-K-N replied:
Fixed it in the latest commit

mkumatag commented on pkg/webhooks/machineset_webhook_test.go:
```diff
@@ -326,98 +420,120 @@ func TestMachineSetCreation(t *testing.T) {
 
 func TestMachineSetUpdate(t *testing.T) {
-	awsClusterID := "aws-cluster"
+	//awsClusterID := "aws-cluster"
 	awsRegion := "region"
-	defaultAWSProviderSpec := &machinev1.AWSMachineProviderConfig{
-		AMI: machinev1.AWSResourceReference{
-			ID: pointer.StringPtr("ami"),
-		},
-		InstanceType:      defaultAWSX86InstanceType,
-		UserDataSecret:    &corev1.LocalObjectReference{Name: defaultUserDataSecret},
-		CredentialsSecret: &corev1.LocalObjectReference{Name: defaultAWSCredentialsSecret},
-		Placement: machinev1.Placement{
-			Region: awsRegion,
-		},
-	}
```
why we are removing this block and also see lot of code commented,

Karthik-K-N replied:
Fixed it in the latest commit

Karthik-K-N commented on pkg/webhooks/machine_webhook.go:
```diff
@@ -1335,3 +1347,96 @@ func validateMachineLifecycleHooks(m, oldM *machinev1.Machine) []error {
+		errs = append(errs, field.Required(field.NewPath("providerSpec", "image"), "providerSpec.image.id or providerSpec.image.name must be provided"))
+	}
+
+	if providerSpec.Network.ID == nil && providerSpec.Network.Name == nil {
```
Machine network validation is done here

Karthik-K-N commented:
/retest

Karthik-K-N commented:
/cc @JoelSpeed @Fedosin

JoelSpeed commented on go.mod:
```diff
@@ -12,13 +12,14 @@ require (
 	github.com/openshift/api v0.0.0-20220325173635-8107b7a38e53
 	github.com/openshift/client-go v0.0.0-20211209144617-7385dd6338e3
 	github.com/openshift/library-go v0.0.0-20220121154930-b7889002d63e
+	github.com/openshift/machine-api-provider-powervs v0.0.2-0.20220404072532-0bd184db2e40
```
This dependency then becomes a dependency of all other providers, which makes bumping the deps a nightmare.

If you want defaulting for the powervs provider spec, it must first be moved into `openshift/api` with the other platforms

Karthik-K-N replied:
Got it, I hope you mean adding the [`PowerVSMachineProviderConfig`](https://github.com/Karthik-K-N/machine-api-provider-powervs/blob/f8d06b4337415daa86e56295c297fef0e615575d/pkg/apis/powervsprovider/v1alpha1/powervsmachineproviderconfig_types.go#L31) to [machine/v1beta1](https://github.com/openshift/api/tree/master/machine/v1beta1)

mkumatag replied:
@Karthik-K-N lets submit a code for the review :)

JoelSpeed replied:
Please make sure to review the API conventions I linked in my previous comment and try to ensure your code meets the requirements before you submit, it will make the process easier

JoelSpeed reviewed (commented):
Codewise I don't have much to comment on, but I don't think we can merge this as is because of the new dependency it introduces.

PowerVS is a new platform for 4.11 right? Has your API had any sort of API review? It would be good to get this done, and, ideally, migrate it into openshift/api

JoelSpeed commented:
I had a conversation with David about the UX here and there are a few validations I would like to see

- Validate the Machine Type is one known to us, if it is not, send a warning to the user that we do not recognise the machine type
- Validate the CPU and Memory restrictions based upon the machine type. The limits are documented by IBM, eg S922 Machines can have up to 22 cores and 4TB of memory, we can validate that and return a good error message to the user at config time rather than at runtime

Karthik-K-N commented:
I have updated this PR to use Power VS apis from opeshift/api repo and This PR is ready for a review

Karthik-K-N commented:
/test unit

JoelSpeed commented on pkg/webhooks/machine_webhook.go:
```diff
@@ -111,6 +147,18 @@ const (
+	powerVSServiceInstance          = "serviceInstance"
+	powerVSNetwork                  = "network"
+	powerVSImage                    = "image"
+	powerVSSystemTypee880           = "e880"
+	powerVSSystemTypee980           = "e980"
```
Nit, I would Capitalise the E to make it clear it's something new
```suggestion
	powerVSSystemTypeE880           = "e880"
	powerVSSystemTypeE980           = "e980"
```

JoelSpeed replied:
This comment still needs resolving

Karthik-K-N replied:
I missed it earlier, Updated it now

JoelSpeed commented on pkg/webhooks/machine_webhook.go:
```diff
@@ -1426,3 +1478,184 @@ func validateAzureDataDisks(machineName string, spec *machinev1.AzureMachineProv
+		providerSpec.ProcessorType = defaultPowerVSProcType
+	}
+	// Doubt: should we make Processors as pointer for easy validation
+	if providerSpec.Processors.IntVal == 0 || providerSpec.Processors.StrVal == "" {
+		providerSpec.Processors = intstr.IntOrString{Type: intstr.String, StrVal: "0.5"}
+	}
```
This depends on the proc type right? We can't set a 0.5 value for some of the proc type values?

JoelSpeed commented on pkg/webhooks/machine_webhook.go:
```diff
@@ -1426,3 +1478,184 @@ func validateAzureDataDisks(machineName string, spec *machinev1.AzureMachineProv
+	if providerSpec.Processors.IntVal == 0 || providerSpec.Processors.StrVal == "" {
+		providerSpec.Processors = intstr.IntOrString{Type: intstr.String, StrVal: "0.5"}
+	}
+	// Doubt: should we make MemoryGiB as pointer for easy validation
```
We don't allow a zero value so we can assume that if it's set to zero, we are going to default it, I think this is fine as is

JoelSpeed commented on pkg/webhooks/machine_webhook.go:
```diff
@@ -1426,3 +1478,184 @@ func validateAzureDataDisks(machineName string, spec *machinev1.AzureMachineProv
+			fmt.Sprintf("%s identifier must be provided", resourceType)))
+	default:
+		errs = append(errs, field.Invalid(parentPath, serviceInstance.Type,
+			fmt.Sprintf("%s identifier is specified as %s but only %s and %s are valid resource identifiers",
```
What about regex type? Why isn't that listed here?

JoelSpeed reviewed (commented):
Not sure how easy this is to do now, but it would be much easier to review if the machinev1 to machinev1beta1 import rename was a separate commit at the beginning of the chain, is that possible?

Karthik-K-N commented:
> Not sure how easy this is to do now, but it would be much easier to review if the machinev1 to machinev1beta1 import rename was a separate commit at the beginning of the chain, is that possible?

I think that can be done, I will give a try, But how about a separate PR with api bump and machinev1beta1 changes.

Karthik-K-N commented:
> > But how about a separate PR with api bump and machinev1beta1 changes.
> 
> That would be even better!

I was able to update it in the same PR, Also i have addressed all your comments.

Karthik-K-N commented:
/hold
hold until respective PR for api update is merged to Power VS mapi:https://github.com/openshift/machine-api-provider-powervs/pull/17

JoelSpeed reviewed (commented):
One nit, but otherwise LGTM

lobziik commented:
/lgtm

JoelSpeed commented:
/approve

Karthik-K-N commented:
/bugzilla refresh

Karthik-K-N commented:
> /approve cancel
> 
> The vendor directory here looks inconsistent, why are there vendor changes without equivalent go.mod and go.sum entries? I can see that we are pulling in API types from the machine-api-provider-powervs repository, that is not allowed. You need to move your API types through to the openshift/api repo before these webhooks can be introduced
> 
> This will have to be done in 4.12

Seems like inconsistencies due to rebase effort, I have udpated it now.

mkumatag commented:
/bugzilla refresh

JoelSpeed reviewed (commented):
Please drop the two vendor commits as they cancel each other out, otherwise LGTM. Will ask for another pair of eyes over it before we merge

Karthik-K-N commented:
> Please drop the two vendor commits as they cancel each other out, otherwise LGTM. Will ask for another pair of eyes over it before we merge

Dropped the vendor commits

Fedosin reviewed (approved):
/lgtm

JoelSpeed commented:
/override ci/prow/e2e-vsphere-serial 

This change doesn't affect vsphere
